Ability to add auto shrinking of connection pool size

jchumnanvech opened the pull request.

haizaar commented:
Thanks for the code. Couple of things are missing though:

 * You shrink will run only once, two minutes after pool init. Tornado's `add_timeout` is a one-time shoot
 * I think the right way to do it is to wait x minutes after the last query. I.e. if the pool is idle - shrink it. I would mark last execution time for each connections and do a periodic timeout in on a `Pool` level that will go over all free connections every `max_conn_idle_time/10` and close idle connections.

Please always write unittests for any features/bugs you submit and update the docs. The code also should pass PEP8 compliance with `pep8 --max-line-length=120`.

jchumnanvech commented:
After every run of auto shrink it re-adds itself back to the ioloop. 

I hesitate about waiting until x minutes after the last query because lets say we put pool size 5 and max to 20. We can anticipate constant connections around 3-4 but every now and then we get high traffic and we need to service 20 concurrent connections. The high traffic period ends and the constant traffic of 3-4 continues for the life time of the app. The connection pool will never shrink back down.

haizaar commented:
My bad, you are right about the timeout.

I see what you are saying. I'm concerned about that fact that you want to shrink free connections without looking at the pool load. Even during high traffic faze, a connection can become free every now and then and there is a chance that it will be shrunk. 

How about:
  * Trying to reuse the same connections. I.e. if your qps is 3 and pool size is 10, then we'll always try to assign the same 3 connection objects to do the work
  * This way it would be meaningful to mark `last_used_time` for each connection. Once it goes over `shrink_delay` we can dispose the connection
  * I would implement it by converting `ConnectionPool.free` to `deque` and use it as a stack. Then least used connections will always be on the left.

What do you think?

jchumnanvech commented:
I added in last_used_time and use that in the calculation of shrink_delay before removing it from the pool and closing the connection. 

I added in a unit test to make sure the pool shrank and made the code pass pep8 inspection. It still complains a little with pep8 about imports not being at the top but it does that on master as well. So no new inspection errors were created.

haizaar commented on momoko/connection.py:
```diff
@@ -56,7 +56,8 @@ def add_free(self, conn):
         if not self.waiting_queue:
             log.debug("No outstanding requests - adding to free pool")
-            self.free.add(conn)
+            conn.last_used_time = datetime.datetime.now()
```
`time.time()` would be faster.

haizaar commented on momoko/connection.py:
```diff
@@ -101,10 +102,23 @@ def abort_waiting_queue(self, error):
+            else:
+                self.free.appendleft(conn)
+                return
+
```
How about:

```
now = time.time()
while len(self.free) > target_size and now - self.free[0].last_used_time >= delay:
    conn = self.free.popleft()
    conn.close()
```

haizaar commented on momoko/connection.py:
```diff
@@ -124,7 +138,11 @@ def __init__(self,
+                 auto_shrink=False,
+                 shrink_period=datetime.timedelta(minutes=2),
+                 shrink_delay=datetime.timedelta(minutes=2)
+                 ):
```
I would put `shrink_delay` before `shrink_period`, since the later one is the "advanced tuning".

haizaar commented on momoko/connection.py:
```diff
@@ -149,6 +167,14 @@ def __init__(self,
         self._last_connect_time = 0
         self._no_conn_availble_error = psycopg2.DatabaseError("No database connection available")
+        self.shrink_period = shrink_period
+        self.shrink_delay = shrink_delay
```
need to convert `shrink_delay` to seconds (because of `time.time()` is used in other places).

jchumnanvech replied:
done

haizaar commented on momoko/connection.py:
```diff
@@ -149,6 +167,14 @@ def __init__(self,
         self._no_conn_availble_error = psycopg2.DatabaseError("No database connection available")
+        self.shrink_period = shrink_period
+        self.shrink_delay = shrink_delay
+        if auto_shrink:
```
I would also record `auto_shrink` value in `self`:

```
self.auto_shrink = auto_shrink
if self.auto_shrink:
    ...
```

haizaar commented:
Looks good to me modulo comments added. Please also update `Pool` docstring to include the new parameters.

Thanks!

haizaar commented on tests.py:
```diff
@@ -722,6 +728,33 @@ def test_partially_connected(self):
+        f3 = db.execute("Select 3")
+        f4 = db.execute("Select 4")
+        f5 = db.execute("Select 5")
+        cursor = yield [f1, f2, f3, f4, f5]
```
I would rename `cursor` to `cursors`.

jchumnanvech commented:
Did all the requests. Added the unit test.

haizaar commented:
Thanks!
